mysys/mf_keycache.c

```diff
@@ -2629,92 +2634,104 @@ static BLOCK_LINK *find_key_block(SIMPLE_KEY_CACHE_CB *keycache,
+              fail_block(block));
+  DBUG_ASSERT((block->length == 0) || fail_block(block));
+  DBUG_ASSERT((block->offset == keycache->key_cache_block_size) ||
+              fail_block(block));
+  DBUG_ASSERT((block->requests > 0) || fail_block(block));
+
+  if (got_length < min_length)
+    block->status|= BLOCK_ERROR;
   else
   {
+    block->status|= BLOCK_READ;
+    block->length= got_length;
     /*
-      This code is executed only by threads that submitted secondary
-      requests. At this point it could happen that the cache block is
-      not yet assigned to the hash_link for the requested file block.
-      But at awake from the wait this should be the case. Unfortunately
-      we cannot assert this here because we do not know the hash_link
-      for the requested file block nor the file and position. So we have
-      to assert this in the caller.
+      Do not set block->offset here. If this block is marked
+      BLOCK_CHANGED later, we want to flush only the modified part. So
+      only a writer may set block->offset down from
+      keycache->key_cache_block_size.
     */
-    KEYCACHE_DBUG_PRINT("read_block",
-                      ("secondary request waiting for new page to be read"));
-    wait_on_queue(&block->wqueue[COND_FOR_REQUESTED], &keycache->cache_lock);
-    KEYCACHE_DBUG_PRINT("read_block",
-                        ("secondary request: new page in cache"));
   }
+  KEYCACHE_DBUG_PRINT("read_block_primary",
+                      ("primary request: new page in cache"));
+  /* Signal that all pending requests for this page now can be processed */
+  release_whole_queue(&block->wqueue[COND_FOR_REQUESTED]);
+
+}
+
+static void read_block_secondary(SIMPLE_KEY_CACHE_CB *keycache,
+                       BLOCK_LINK *block)
+{
+  KEYCACHE_THREAD_TRACE("read_block_secondary");
+
+  /*
+    This code is executed only by threads that submitted secondary
+    requests. At this point it could happen that the cache block is
+    not yet assigned to the hash_link for the requested file block.
+    But at awake from the wait this should be the case. Unfortunately
+    we cannot assert this here because we do not know the hash_link
+    for the requested file block nor the file and position. So we have
+    to assert this in the caller.
+  */
+  KEYCACHE_DBUG_PRINT("read_block_secondary",
+                    ("secondary request waiting for new page to be read"));
+  wait_on_queue(&block->wqueue[COND_FOR_REQUESTED], &keycache->cache_lock);
+  KEYCACHE_DBUG_PRINT("read_block_secondary",
+                      ("secondary request: new page in cache"));
 }
 
+static inline void read_block(SIMPLE_KEY_CACHE_CB *keycache,
+                       BLOCK_LINK *block, uint read_length,
+                       uint min_length, my_bool primary)
+{
+  if (primary)
+    read_block_primary(keycache, block, read_length, min_length);
+  else
+    read_block_secondary(keycache, block);
+}
```

Comment:
Compared to other changes, I can't immediately see justification to ```read_block()``` split. I'd understand it if you would use them directly.

So did you want to assign some TSAN attributes? If so, what kind of attributes and why?

Reply:
I just re-read callers. They do duplicate checks for ```page_st```, so in fact they can call primary/secondary directly without big code changes.